CLI "show running configuration" HLD

bhaveshdell opened the pull request:
Management framework- CLI "show running conifguration" design doc.

bhaveshdell committed:
Management framework- CLI "show running conifguration" Document.

bhaveshdell committed:
Rename the files.

bhaveshdell committed:
Update image file name.

bhaveshdell committed:
Update command and view rendering sections.

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
```
I'm going to focus my review on requirements - I don't understand the rest of it well enough to add any value there.

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
+#### About this Manual
+
+This document states the functional requirements and high level design of the new feature to support "show running-configuration" command.
```
Shouldn't it be "show running-config" (per Cisco)?

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
+This document states the functional requirements and high level design of the new feature to support "show running-configuration" command.
+
+#### Scope
```
Make clear that the scope is limited to Klish and not other CLIs in the system (e.g. Click, vtysh). It is also limited to configuration that is found in the CONFIG_DB.

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+For this command to show the entire running configuration of the system,
+all apps should support configuration through CLI interface.
+
+Any configuration done using GNMI or REST or other management
```
gNMI not GNMI - throughout

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+all apps should support configuration through CLI interface.
+
+Any configuration done using GNMI or REST or other management
+interface which is stored and has a corresponding CLI
```
"which is stored and has a corresponding CLI command" -> "which is stored in CONFIG_DB and has a corresponding Klish CLI command"

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+interface which is stored and has a corresponding CLI
+command will be displayed as part of this new command.
+
+The command output can be copied and pasted to be used as
```
"and should arrive at the same running configuration as the source session"

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+The order of commands displayed in the "show running-configuration" will be
+in accordance to the order of dependency between CLI views.
+
+If configuration default values are stored, they will be shown as
```
"default values are stored" - what does this mean? They are in the CONFIG_DB? IOW, in-code defaults are not shown?

Note that we should have 2 options: -
1) show running-config       ; show non-default
2) show running-config all   ; Include defaults

See here - https://www.cisco.com/c/m/en_us/techdoc/dc/reference/cli/nxos/commands/fund/show-running-config.html

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+part of "show running-configuration".
+
+The infrastructure will provide functionality to filter and display
+the configuration based on the CLI views.
```
Pls explain what a view is - I think this is a command mode, correct?

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+The infrastructure will provide functionality to filter and display
+the configuration based on the CLI views.
+
+A plugin in form of callback will be provided for apps to do their own database retrieval and
```
I think this is a design statement rather than a requirement. The requirement should be something like "The infrastructure automates as much of the show running-config output generation as possible from the developer-provided information. Where more developer work is required, an easy-to-use plug-in infrastructure is provided".

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+A plugin in form of callback will be provided for apps to do their own database retrieval and
+rendering. Apps will return the list of commands to the infra which will then be
+added to the final list of commands to be displayed.
+
```
I would also suggest: -

- It should be possible to pipe the output to a file or script, and to be able to copy this on/off box.
- Pagination options.
- The output should appear in show techsupport
- Don't show protected information (e.g. passwords/keys) in clear text
- Cisco has commands around copying between the running-config and the startup-config (and other files) - should this be here also?

https://www.cisco.com/c/en/us/td/docs/switches/datacenter/nexus9000/sw/7-x/fundamentals/configuration/guide/b_Cisco_Nexus_9000_Series_NX-OS_Fundamentals_Configuration_Guide_7x/b_Cisco_Nexus_9000_Series_NX-OS_Fundamentals_Configuration_Guide_7x_chapter_01001.html

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
+System state information should not be added to this command.
+
+The ordering between commands in the same view or between views is not supported. Hence some set of configuration will not work for copy-paste operation.
```
What does this mean? This seems serious. We need to be able to generate/paste successfully in all cases.

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+Example:
+
+sonic(conf)\# interface PortChannel 5  
+sonic(conf-if-po5)\#show configuration  
```
This is not a Cisco thing, right? Cisco instead seems to have "show running-config <feature_name>" commands. Wouldn't this be more industry-standard?

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
+For these exceptions two approaches are provided to the developer:
+
+a.  Developer to provide a python callback function or jinja template. The infra will call this
```
Python and Jinja - throughout

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+    function with the required DB tables and keys. The
+    function will return a list of commands. This list will be
+    incorporated into the final command output.
+    This can also be done using a jinja template. Python callback is
```
So why offer the Jinja path?

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
+The automation infra retrieves data from DB using the REST
+interface and sonic yang model. The developer of CLI has to provide in
+the cli xml, the DB table and attribute information related to
```
CLI and XML - throughout

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+    of commands to the infra to be displayed in the final output.
+
+The automation infra retrieves data from DB using the REST
+interface and sonic yang model. The developer of CLI has to provide in
```
Remove the "in"

Also YANG not yang - throughout

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+
+## **SONiC Management Framework CLI 'Show Running Configuration'**
```
Why haven't you used the HLD template for this doc?

ben-gale commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,495 @@
+1.  Compile mgmt-framework package and install.
+2.  Verify the command output formats.
+
+#### 5 Unit Test and Automation
```
Please fill out

bhaveshdell committed:
Address review comments.
Update to developer guide sections.
Add new XLM tag information.
Add more examples for development.

bhaveshdell committed:
Revise and update.

sachinholla commented on doc/mgmt/Show_Running_Configuration_HLD.md:
```diff
@@ -0,0 +1,654 @@
+                    The mapping is to be provided in file CLI/acioner/show_config_data.py
+                    In most cases, the command rendering is short and can be done efficiently in Python. 
+                    A Jinja template has overhead of loading which increases linearly with file size.
+                    For command rendering it is recommended to implement it in Python. 
```
Please mention that multiple command lines should be joined using semicolon delimiter.

dr412113 commented on doc/mgmt/show_running_cfg.md:
```diff
@@ -0,0 +1,495 @@
+The order of commands displayed in the "show running-configuration" will be
+in accordance to the order of dependency between CLI views.
+
+If configuration default values are stored, they will be shown as
```
Why do we want to show all default configs under show running-config?
It will not be readable and i believe it is not inline with industry standards.

bhaveshdell committed:
Update users guide with section on command formats.
Add section on command order.
Add Table of Contents.
